components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,394 @@
+            }
+            Version::V2 => {
+                file_dict.merge_from_bytes(content)?;
+                // parse the remained records
+                while !remained.is_empty() {
+                    // If the parsing get errors, manual intervention is required to recover.
+                    let (used_size, file_name, mode) = Self::parse_next_record(remained)?;
+                    remained.consume(used_size);
+                    match mode {
+                        LogRecord::INSERT(info) => {
+                            file_dict.files.insert(file_name, info);
+                        }
+                        LogRecord::REMOVE => {
+                            let original = file_dict.files.remove(&file_name);
+                            if original.is_none() {
+                                return Err(box_err!(
+                                    "Try to recovery from log file but remove a null entry, file name: {}",
+                                    file_name
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&mut self, name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&mut self, name: &str) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::REMOVE)?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Replace existed file entry with new file information.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn replace(&mut self, src_name: &str, dst_name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes1 = Self::convert_record_to_bytes(src_name, LogRecord::REMOVE)?;
+            let bytes2 = Self::convert_record_to_bytes(dst_name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes1)?;
```

Comment:
These 2 write operations are not atomic. I think a `RENAME` operation is required.

Reply:
I prefer to keep IO atomic, but keep append insert operation first. So that if the remove operation fails to write, it is guaranteed that file information will not be lost